### ISSUE TITLE
Unhandled exception inside Interop+OpenSsl.AlpnServerSelectCallback causing the process crash

### ISSUE THREAD

**Issue by VasiliyNovikov** [dotnet/runtime#81588]
### Description

Unhandled exception which causes process to crash

```
System.InvalidOperationException: Handle is not initialized.
Stack trace:
  in System.Void System.ThrowHelper.ThrowInvalidOperationException_HandleIsNotInitialized()
  in System.Runtime.InteropServices.GCHandle System.Runtime.InteropServices.GCHandle.FromIntPtr(System.IntPtr)
  in System.Int32 Interop+OpenSsl.AlpnServerSelectCallback(System.IntPtr, System.Byte**, System.Byte*, System.Byte*, System.UInt32, System.IntPtr)
```

### Reproduction Steps

It reproduces sometimes (but extremely rarely) in ASP.NET Core app running on Ubuntu 18:04 Server after we migrated from .NET 6 to .NET 7
It uses Kestrel + SSL certificate for HTTPS

### Expected behavior

-

### Actual behavior

Process crashes

### Regression?

Maybe .NET 7.0 regression

### Configuration

uname -a
Linux tll-netem-1 5.4.0-87-lowlatency #98~18.04.1-Ubuntu SMP PREEMPT Wed Sep 22 13:43:18 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

dotnet --info
.NET SDK:
 Version:   7.0.102
 Commit:    4bbdd14480

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.102/

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  5.0.408 [/usr/share/dotnet/sdk]
  6.0.405 [/usr/share/dotnet/sdk]
  7.0.102 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.13 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

**Comment by rzikm** [comment]
we should look at it in 8.0.

@VasiliyNovikov crash dumps would be very useful for debugging this. Do you have some available?

**Comment by VasiliyNovikov** [comment]
Unfortunately no, it was reproduced in prod environment only twice. I'll try to configure dump collection so on next repro I could get one

**Comment by janvorli** [comment]
@VasiliyNovikov is the issue that the unhandled exception occurs at all or that it aborts the process while you'd expect it to be handled somewhere up the call chain?

**Comment by janvorli** [comment]
The thing is that unhandled exceptions abort the process by their nature. So that's why I was asking if you have expected that exception to get handled somewhere up the call chain.
The important thing to mention is that on Unix, exceptions cannot be propagated from callbacks that are called by native code.

**Comment by VasiliyNovikov** [comment]
There seems to be no call chain where I can handle it. That the only stack trace I've got

**Comment by wfurt** [comment]
https://github.com/dotnet/runtime/blob/8300e381e41bdf013c5dc697e3e9c15f68e8f249/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs#L679-L680

It is called directly from OpenSSL. We should add guard agains null pointer there.

**Comment by wfurt** [comment]
Btw what OpenSSL version do you have @VasiliyNovikov?

**Comment by VasiliyNovikov** [comment]
OpenSSL 1.1.1  11 Sep 2018

**Comment by VasiliyNovikov** [comment]
I configured crash dump collection, so next time it happens I hope to get one

**Comment by VasiliyNovikov** [comment]
I checked the all occurrences and found a pattern - it happens every 2nd day around the same time. Since each time it crashes it almost immediately restarts - I guess it just always crashes after 2 days of running. After I configured crash dump collection I manually restarted it so let's see if next occurrence will be in 2 days

**Comment by wfurt** [comment]
reopening for possible 7.x backport

**Comment by VasiliyNovikov** [comment]
> I checked the all occurrences and found a pattern - it happens every 2nd day around the same time. Since each time it crashes it almost immediately restarts - I guess it just always crashes after 2 days of running. After I configured crash dump collection I manually restarted it so let's see if next occurrence will be in 2 days

My hypothesis was wrong about it being crashed after 2 days of running. It seems to crash at some specific time each 2 days no-matter if I restarted the process

I collected a crash dump,  so what would be a good way to send it to you (if you still need it)? I might contain some sensitive data, so it should be some non-publically available location

CC @rzikm , @wfurt

**Comment by wfurt** [comment]
thanks for update @VasiliyNovikov I don't know if there is possibility for you to consume daily 8.0 builds and verify the fix:  https://github.com/dotnet/installer (https://github.com/dotnet/runtime/pull/81629 missed the preview1) 
I'll try to get approval for 7.0.x backport because of the impact.

**Comment by VasiliyNovikov** [comment]
> thanks for update @VasiliyNovikov I don't know if there is possibility for you to consume daily 8.0 builds and verify the fix: https://github.com/dotnet/installer (#81629 missed the preview1) I'll try to get approval for 7.0.x backport because of the impact.

It is not so easy since the issue is reproduced only on one production machine

**Comment by VasiliyNovikov** [comment]
So do you still need a crash dump? I have one

**Comment by wfurt** [comment]
yes, we do not know the root cause - the fix only prevents the uncatchable exception that aborts the process. You can send the link to my GH email and possibly add @rzikm as well.

**Comment by karelz** [comment]
@VasiliyNovikov few questions for you:
- How many machines run without any problem, not hitting this issue ... it will help us understand the impact on your larger service of the 1 machine.
- One machine problems are tricky. In extreme cases it can be HW or configuration errors. This does not look like that, but it is not certain. Are you in position to take the bad machine out of rotation? Or reimage it? Or otherwise "clean it"?
- BEFORE you start reimaging the machine. Are you in position to test private bits from servicing branch on it with the fix @wfurt prepared?

**Comment by rzikm** [comment]
It would be helpful if you could test out @wfurt's fix:

- Publish your app as self contained (`dotnet publish --self-contained`, **not** single-file)
- Replace System.Net.Security.dll with [this one (cherry-picked from the CI pipeline)](https://github.com/dotnet/runtime/files/10797123/System.Net.Security.zip)
- Try to reproduce the issue.

**Comment by VasiliyNovikov** [comment]
@karelz,
- We have only one machine out of ~12 which reproduces the issue
- Reimaging the machine can be an option if everything else won't help but it might take some time
- I'll try to test the fix prepared by @rzikm

@rzikm, I'll try your fix

**Comment by karelz** [comment]
@VasiliyNovikov any update? Did you get a chance to try it? (hoping it was not too difficult)

**Comment by VasiliyNovikov** [comment]
@karelz , I tried it and it seems it fixes the problem. At least I didn't have crashes for almost a week. Before it crashed every 2 days

**Comment by karelz** [comment]
Do you mean you tried the @rzikm's step to include the change in your app?

**Comment by VasiliyNovikov** [comment]
@karelz , yes

**Comment by karelz** [comment]
Great, thanks!
Now we need to understand better why it is happening. As explained in the PR #81797, this is a band-aid fix, we do not want to put in servicing, unless we either understand better the underlying problem, or we have more customers hitting it.

@VasiliyNovikov would you be able to help us debug why the problem is happening?
If not, the other alternative is to try to re-image the machine to see if it is just some setup fluke ...

**Comment by rzikm** [comment]
I checked it briefly but did not uncover the root cause yet, I would need to set it up in a way I can examine openssl structures. Since the problem seems to go away with the bandaid fix, we will postpone deeper investigation until we are done with more urgent stuff.

**Comment by rzikm** [comment]
@larspars Is the repro case similar? Or do you have a simpler one you could share with us?
Also, can you check if the fix from the PR (https://github.com/dotnet/runtime/issues/81588#issuecomment-1439019813) makes the crash go away?

**Comment by karelz** [comment]
@larspars yes please, if you can try it out, that would be super useful to us and probably the tipping point to service 7.0 with it. Thanks!

**Comment by larspars** [comment]
@karelz I've deployed with it now, no problems yet (it's run for an hour). Prior to this, the server crashed 6 times in 3 days after upgrading to net 7. So if this fix runs without problems for some days it should strengthen the argument that the fix is helpful. I'll come back here with an update when it has run for longer.

**Comment by rzikm** [comment]
@wfurt s bandaind fix has been merged to 7.0.5.

**Comment by karelz** [comment]
Fixed in main (8.0) in PR #81629 and in 7.0.5 in PR #81797.
